modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -0,0 +1,718 @@
+        }
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     */
+    public void runJdbcDml(IgniteBiTuple<List<String>, Boolean> qrysInfo) {
+        assumeFalse("There is no 'local query' parameter for JDBC queries", loc);
+
+        executeDml(qrysInfo, (cmds) -> {
+            try (
+                Connection conn = GridTestUtils.connect(queryNode(), null);
+                Statement stmt = conn.createStatement()
+            ) {
+                for (String cmd : cmds)
+                    stmt.execute(cmd);
+            }
+            catch (SQLException e) {
+                new RuntimeException(e);
+            }
+        });
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     */
+    public void runSqlFieldsDml(IgniteBiTuple<List<String>, Boolean> qrysInfo) {
+        executeDml(qrysInfo, (cmds) -> {
+            IgniteCache<Integer, String> cache = queryNode().getOrCreateCache("A");
+
+            cmds.forEach(cmd -> cache.query(new SqlFieldsQuery(cmd).setLocal(loc)));
+        });
+    }
+
+    /**
+     * @param qrysInfo DML commands info (queries, simple query flag).
+     * @param task Task to be executed.
+     */
+    public void executeDml(IgniteBiTuple<List<String>, Boolean> qrysInfo, Consumer<List<String>> task) {
+        assumeFalse("There is no lazy mode for DML operations", !isFullyFetched);
+
+        List<String> cmds = qrysInfo.get1();
+
+        task.accept(cmds);
+
+        checkSqlPlanHistoryDml(3, qrysInfo.get2());
+    }
+
+    /** Returns current SQL plan history on the query node. */
+    public Map<SqlPlan, Long> getSqlPlanHistory() {
+        return queryNode().context().query().runningQueryManager().planHistoryTracker().sqlPlanHistory();
+    }
+
+    /**
+     * Checks SQL plan history for select queries.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     */
+    public void checkSqlPlanHistory(int size) {
+        if (!isReducePhase && isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME)
+            size = 0;
+
+        executeHistoryCheck(size, plans -> plans);
+    }
+
+    /**
+     * Checks SQL plan history for DML operations.
+     *
+     * @param size Number of SQL plan entries expected to be in the history.
+     * @param isSimpleQry Simple query flag.
+     */
+    public void checkSqlPlanHistoryDml(int size, boolean isSimpleQry) {
+        executeHistoryCheck(size, plans -> {
+            if (sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) {
+                String check;
+
+                if (isSimpleQry)
+                    check = "no SELECT queries have been executed.";
+                else
+                    check = "the following " + (loc ? "local " : "") + "query has been executed:";
+
+                plans = plans.entrySet().stream()
+                    .filter(e -> e.getKey().plan().contains(check))
```

Comment:
Why do we filter these queries? Do we have another queries in the view?

Reply:
Yes, in some cases we may also have plans for those select queries that are executed as part of the DML operation (with the H2 engine only).